File: src/main/java/com/micronaut/bug/config/JwtAuthenticationFilter.java
package com.micronaut.bug.config;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.extern.slf4j.Slf4j;
import org.slf4j.MDC;
import org.springframework.core.Ordered;
import org.springframework.core.annotation.Order;
import org.springframework.http.HttpHeaders;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import java.io.IOException;

import static com.micronaut.bug.config.ObservationConfig.TARGET_ID;

@Slf4j
@Component
@Order(Ordered.HIGHEST_PRECEDENCE + 1)
public class JwtAuthenticationFilter extends OncePerRequestFilter {

    @Override
    protected void doFilterInternal(HttpServletRequest request,
                                    HttpServletResponse response,
                                    FilterChain filterChain) throws ServletException, IOException {

        String authHeader = request.getHeader(HttpHeaders.AUTHORIZATION);
        User user = null;

        if (authHeader != null && authHeader.startsWith("Bearer ")) {
            String token = authHeader.substring(7);
            user = parseToken(token);
        }

        try {
            if (user != null) {
                // Привязываем пользователя к области видимости текущего вызова
                // Все, что внутри .run(), будет видеть этого юзера через SecurityContext.CURRENT_USER.get()
                MDC.put(TARGET_ID, user.id());
                final User finalUser = user;
                ScopedValue.where(SecurityContext.CURRENT_USER, finalUser)
                    .run(() -> {
                        try {
                            filterChain.doFilter(request, response);
                        } catch (Exception e) {
                            throw new RuntimeException(e);
                        }
                    });
            } else {
                // Если пользователя нет, просто идем дальше по цепочке
                filterChain.doFilter(request, response);
            }
        } finally {
            MDC.remove(TARGET_ID);
        }
    }

    private User parseToken(String token) {
        // Ваша логика парсинга JWT
        return new User("123", "ivan_ivanov", "ADMIN");
    }
}

File: src/main/kotlin/com/micronaut/bug/config/JwtAuthenticationFilter.kt
package com.micronaut.bug.config

import com.micronaut.bug.trace.NanoTraceFilter.Companion.MDC_USER_ID
import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.slf4j.MDC
import org.springframework.core.Ordered
import org.springframework.core.annotation.Order
import org.springframework.http.HttpHeaders
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component
@Order(Ordered.HIGHEST_PRECEDENCE + 1)
class JwtAuthenticationFilter : OncePerRequestFilter() {

    private val log = KotlinLogging.logger {}

    override fun doFilterInternal(
        rq: HttpServletRequest,
        rs: HttpServletResponse,
        chain: FilterChain
    ) {
        val authHeader = rq.getHeader(HttpHeaders.AUTHORIZATION)
        val user = parseToken("")
//      val user = if (authHeader != null && authHeader.startsWith(BEARER_PREFIX)) {
//            val token = authHeader.substring(BEARER_PREFIX.length)
//            parseToken(token)
//        } else {
//            null
//        }

        try {
            if (user != null) {
                MDC.put(MDC_USER_ID, user.id)
                SecurityContext.set(user)
            }
            chain.doFilter(rq, rs)
        } finally {
            SecurityContext.clear()
        }
    }

    private fun parseToken(token: String): User {
        // Ваша логика парсинга JWT
        return User("123", "ivan_ivanov", "ADMIN")
    }

    companion object {
        private const val BEARER_PREFIX = "Bearer "
    }
}
